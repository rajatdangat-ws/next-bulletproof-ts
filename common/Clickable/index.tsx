/**
 *
 * Clickable
 *
 */

import React, { ReactEventHandler } from "react";
import styled from "styled-components";
import { T } from "@common";

interface Props {
  onClick: ReactEventHandler;
  textId: string;
}

const StyledClickable = styled.div`
  color: #1890ff;
  &:hover {
    cursor: pointer;
  }
`;

const Clickable: React.FC<Props> = ({ onClick, textId }) => {
  return (
    <StyledClickable data-testid="clickable" onClick={onClick}>
      {textId && <T id={textId} />}
    </StyledClickable>
  );
};

export default Clickable;
