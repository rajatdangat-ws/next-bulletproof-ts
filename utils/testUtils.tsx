import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { IntlProvider } from "react-intl";
import { Provider as ReduxProvider } from "react-redux";
import { store } from "../store";
import messages from "../translations/en.json";

/**
 * Renders the passed in components or tree with all the providers.
 * To configure just add more providers and use the function in tests to render with providers
 * @param children, all the components in the composition
 * @returns React.FC, renders the DOM with all the providers
 */
export const WithAllProviders: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <IntlProvider messages={messages} locale="en" defaultLocale="en">
        {children}
      </IntlProvider>
    </ReduxProvider>
  );
};

/**
 * Generic render function that renders all the listed components with all the providers
 * @param ui The ReactElement (ie. Component) to render
 * @param options extra render options
 * @returns returns the RenderResult with all the things that react-testing library provides
 */
export const renderForTesting = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithAllProviders, ...options });

/**
 * Stops execution in tests for a specified milliseconds
 * @param ms time in milliseconds, how long to stop
 * @returns returns a new Promise to hang
 */
export const timeout = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Generate an API Response that can be used to mock API Responses
 * @param ok boolean specifing will the response resolve or reject
 * @param data the data that is to be sent back
 * @returns APIResponse consisting ok and data
 */
export function apiResponseGenerator<T>(ok: boolean, data: T) {
  return {
    ok,
    data,
  };
}

export * from "@testing-library/react";
export { renderForTesting as render };
