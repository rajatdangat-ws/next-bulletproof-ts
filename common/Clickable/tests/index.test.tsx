/**
 *
 * Tests for Clickable
 *
 */

import React from "react";
import { fireEvent } from "@testing-library/dom";
import { render } from "@utils/testUtils";
import Clickable from "../index";

describe("<Clickable /> component tests", () => {
  let clickSpy = jest.fn();
  beforeEach(() => {
    clickSpy = jest.fn();
  });

  it("should render and match the snapshot", () => {
    const { baseElement } = render(<Clickable onClick={clickSpy} textId="repo_list" />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should contain 1 Clickable component", () => {
    const { getAllByTestId } = render(<Clickable onClick={clickSpy} textId="repo_list" />);
    expect(getAllByTestId("clickable").length).toBe(1);
  });

  it("should contain render the text according to the textId", () => {
    const { getAllByText } = render(<Clickable onClick={clickSpy} textId="repo_list" />);
    expect(getAllByText(/Repository List/).length).toBe(1);
  });

  it("should call the prop onClick when the clickable component is clicked", () => {
    const { getAllByText, queryByText } = render(
      <Clickable onClick={clickSpy} textId="repo_list" />
    );

    expect(getAllByText(/Repository List/).length).toBe(1);

    const elementToClick = queryByText(/Repository List/);
    if (elementToClick) {
      fireEvent.click(elementToClick);
      expect(clickSpy).toBeCalled();
    }
  });
});
