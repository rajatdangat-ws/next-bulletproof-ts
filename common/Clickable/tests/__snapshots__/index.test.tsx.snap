// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clickable /> component tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="Clickable__StyledClickable-sc-wshbsq-0 drTgFv"
      data-testid="clickable"
    >
      <p
        class="T__StyledText-sc-apdcd0-0 fjJkvM"
        data-testid="t"
      >
        Repository List
      </p>
    </div>
  </div>
</body>
`;
